### PR TITLE
[ADAG] Support multi-args and kwargs in Accelerate DAG

### LINKED ISSUE #42793: [core][dags] Support multiple args and kwargs in accelerated DAGs
### Description

Right now the experimental accelerated DAGs feature only supports a single argument and output per task. Ideally we want to be able to pass/return multiple. This can be done by serializing the argument list/dict and passing it through the channel, instead of [assuming and serializing only one value](https://github.com/ray-project/ray/blob/master/python/ray/experimental/channel.py#L118).

We should also support no-args.

### Use case

_No response_

### PR DESCRIPTION
**Pull request by ruisearch42**
## Why are these changes needed?
Currently accelerated DAG only supports a single argument per task. We would like to add general support for multi-args and kwargs.

This PR adds multi-arg and kwarg support by serializing all positional args and kwargs and passing it through the channel. When the channel is read at runtime, the individual args are extracted first before passing to the consuming tasks.

## Related issue number
Closes #42793

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/dag/compiled_dag_node.py [hunk 1]
```diff
@@ -798,24 +882,17 @@ def execute(
...
-            raise NotImplementedError("Compiled DAGs do not support kwargs")
-
         if self._enable_asyncio:
             raise ValueError("Use execute_async if enable_asyncio=True")
 
         self._get_or_compile()
 
-        inp = args[0]
+        inp = (args, kwargs)
         if self.input_wrapper_fn is not None:
```
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:896]: TODO: double check this (TorchTensorWrapper) works properly.

File: python/ray/dag/compiled_dag_node.py [hunk 2]
```diff
@@ -458,6 +527,22 @@ def _preprocess(self) -> None:
...
+
+                # If the upstream node is an InputAttributeNode, treat the
+                # DAG's input node as the actual upstream node
+                if isinstance(upstream_node.dag_node, InputAttributeNode):
+                    assert (
+                        "dag_input_node"
+                        in upstream_node.dag_node.get_other_args_to_resolve()
+                    )
+                    dag_input_node_index = self.dag_node_to_idx[
+                        upstream_node.dag_node.get_other_args_to_resolve()[
```
Comment (stephanie-wang) [inline python/ray/dag/compiled_dag_node.py:539]: Suggest separating this out into a helper method on the `InputAttributeNode` class that returns the corresponding `InputNode`.

File: python/ray/dag/compiled_dag_node.py (hunk 1 above)
Comment (stephanie-wang) [reply python/ray/dag/compiled_dag_node.py:897]: Ah I believe this actually should've been removed in #45092, feel free to remove it.

File: python/ray/dag/compiled_dag_node.py [hunk 3]
```diff
@@ -284,7 +308,46 @@ def __init__(
 
         self.resolved_inputs: List[Union[Any, ChannelInterface]] = []
         self.input_channels: List[ChannelInterface] = []
-        self.input_channel_idxs: List[int] = []
+        # The inner list is indices of resolved_args that correspond
```
Comment (stephanie-wang) [inline python/ray/dag/compiled_dag_node.py:311]: The logic here is a little bit hard to follow. I wonder if we can change to something more like this?

```python
# represents an arg read from a channel.
# (index into self.input_channels, index into self.resolved_inputs, arg / kwarg / None to use the value directly)
ChannelInput = Tuple[int, int, Optional[Union[int, str]]]

self.channel_inputs: List[ChannelInput]
```

It would also be great to consolidate the logic to build the channel inputs into the ExecutableTask constructor, and move input resolution into a method. That way we can make these "private" (or as private as Python lets us :) )

File: python/ray/dag/tests/experimental/test_accelerated_dag.py [hunk 4]
```diff
@@ -175,6 +181,129 @@ def test_regular_args(ray_start_regular):
...
 
 
+def test_multi_args_basic(ray_start_regular):
+    a1 = Actor.remote(0)
+    a2 = Actor.remote(0)
+    a3 = Actor.remote(0)
+    with InputNode() as i:
+        branch1 = a1.inc.bind(i[0])
+        branch2 = a2.inc.bind(i[1])
+        dag = a3.inc_two.bind(branch1, branch2)
```
Comment (stephanie-wang) [inline python/ray/dag/tests/experimental/test_accelerated_dag.py:191]: Nice tests! However, could you change the tested method to concatenate the results or something like that? When we use addition, we can't tell from the end result if the args were passed in the correct order.

**Review by stephanie-wang — COMMENTED** [review]
Nice! The logic looks good to me; I left some suggestions for cleanup.

File: python/ray/dag/compiled_dag_node.py (hunk 1 above)
Comment (ruisearch42) [reply python/ray/dag/compiled_dag_node.py:897]: Thanks. Removed.

File: python/ray/dag/compiled_dag_node.py (hunk 2 above)
Comment (ruisearch42) [reply python/ray/dag/compiled_dag_node.py:539]: The `InputNode` is already known so we don't need it from `InputAttributeNode`. This was a sanity check that can probably be removed. I cleaned up the code.

File: python/ray/dag/compiled_dag_node.py (hunk 3 above)
Comment (ruisearch42) [reply python/ray/dag/compiled_dag_node.py:311]: Good suggestion. I refactored the code and introduced `ExecutableTaskInput`.

File: python/ray/dag/compiled_dag_node.py [hunk 5]
```diff
@@ -243,6 +235,85 @@ def __str__(self) -> str:
...
+
+    def __init__(
+        self,
+        input_variant: Union[ChannelInterface, DAGInputAdapter, Any],
+        channel_idx: Optional[int],
+    ):
+        self.input_variant = input_variant
+        self.channel_idx = channel_idx
+
+    def resolve(self, channel_results):
```
Comment (stephanie-wang) [inline python/ray/dag/compiled_dag_node.py:300]: Can you add the type hint for channel_results?

File: python/ray/dag/compiled_dag_node.py [hunk 6]
```diff
@@ -243,6 +235,85 @@ def __str__(self) -> str:
 """
 
 
+@DeveloperAPI
```
Comment (stephanie-wang) [inline python/ray/dag/compiled_dag_node.py:238]: I think these should not be developer APIs. Prepending the class name with underscore will allow you to not specify an API decorator.

**Review by stephanie-wang — APPROVED** [review]
